### ISSUE TITLE
Issue with triplet loss

### ISSUE THREAD

**Issue by cattaneod** [mikacuy/pointnetvlad#2]
I think your implementation of best_pos_distance (which is used for every loss in your code) should use reduce_max instead of reduce_min (following the original hard triplet loss paper).
I re-implemented your code in PyTorch, and with max pos_distance i got an average top 1% recall on oxford baseline (no refine) of 84%

**Comment by mikacuy** [comment]
Hi,

Thank you for your interest in our work. The definition of triplet loss is shown below:
![screen shot 2018-06-05 at 8 03 17 pm](https://user-images.githubusercontent.com/9430036/40974946-e2d0f22e-68fb-11e8-89ab-c8f3ff2cf51f.png)
where a, p, n are the anchor and corresponding positive and negative, respectively. Simply following this definition, there is no need for the usage of "reduce_min" or "reduce_max". Either case, it is still the triplet loss

As stated under Implementation Details in our supplementary material, we just found that if we took the best/closest among two randomly sampled positives for each tuple, we found the training to be more stable which is why we used "reduced_min", ie taking the closest positive. The original NetVLAD paper also takes the best positive in each training tuple for their loss function because of their weakly supervised approach. Using "reduce_max" is also possible, both are correct, the interpretation of using "reduce_max" is taking the **HARDEST** positive in the training tuple instead of the best. So this can potentially be used to further improve the model because it will result in a higher loss value. If you are able to improve the results by further training with "reduce_max" then that's great!

The implementation of using "reduce_max" is also in the code I uploaded, it was just commented out because I found some stability issue initially.

Regards,
Mika

**Comment by cattaneod** [comment]
Thanks for your answer,
I observed that with reduce_min, the loss in later epochs is often zero (around half of the time).
Just for personal curiosity, i will try to train again my implementation with the closest positive match (i only trained it with the hardest positive).

**Comment by mikacuy** [comment]
Yes, reduce_min will eventually give a loss of 0 in the later epochs. I suspect reduce_max eventually will also be the same. I think that the performance can also possibly improve if you further train with reduce_max. 

Actually, currently the best model we have achieved is also 84% for top 1% (baseline). We will upload this model soon.

**Comment by SJTUsuperxu** [comment]
@catta202000 Hi, I also wan to port the tensorflow based code to pytorch. I was wondering how you implement VLAD. In my code, I define a conv layer and parameters for cluster features(size[Num_clusters, Dim_features])．And the cluster features are initialized from random. When I train the whole net, the loss behaves strangely: decreases to zero rapidly. Could you please share your implementation of VLAD with me?? Thank you very much.

**Comment by cattaneod** [comment]
@SJTUsuperxu Hi, my implementation is based on [LOUPE tensorflow toolbox](https://github.com/antoine77340/LOUPE), you can find my PyTorch implementation here: [Link](https://gist.github.com/catta202000/cbf5cae9199ccd72adddef452358df3b)

I think i will release my whole PointNetVlad Pytorch implementation soon.

**Comment by SJTUsuperxu** [comment]
@catta202000 @mikacuy    Thanks a lot

**Comment by SJTUsuperxu** [comment]
@catta202000  Now I can get access to this link.. I will look into it in details. Thanks.

**Comment by catta202000** [comment]
Try now, i can open it without problem.

**Comment by SJTUsuperxu** [comment]
I can access it now. Thank you very much.-------- 原始邮件 --------主题：Re: [mikacuy/pointnetvlad] Issue with triplet loss (#2)发件人：catta202000 收件人：mikacuy/pointnetvlad 抄送：Jingwei Xu ,Mention Try now, i can open it without problem.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
{"@context":"http://schema.org","@type":"EmailMessage","potentialAction":{"@type":"ViewAction","target":"https://github.com/mikacuy/pointnetvlad/issues/2#issuecomment-395726277","url":"https://github.com/mikacuy/pointnetvlad/issues/2#issuecomment-395726277","name":"View Issue"},"description":"View this Issue on GitHub","publisher":{"@type":"Organization","name":"GitHub","url":"https://github.com"}}
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/mikacuy/pointnetvlad","title":"mikacuy/pointnetvlad","subtitle":"GitHub repository","main_image_url":"https://assets-cdn.github.com/images/email/message_cards/header.png","avatar_image_url":"https://assets-cdn.github.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/mikacuy/pointnetvlad"}},"updates":{"snippets":[{"icon":"PERSON","message":"@catta202000 in #2: Try now, i can open it without problem."}],"action":{"name":"View Issue","url":"https://github.com/mikacuy/pointnetvlad/issues/2#issuecomment-395726277"}}}
{
"@type": "MessageCard",
"@context": "http://schema.org/extensions",
"hideOriginalBody": "false",
"originator": "AF6C5A86-E920-430C-9C59-A73278B5EFEB",
"title": "Re: [mikacuy/pointnetvlad] Issue with triplet loss (#2)",
"sections": [
{
"text": "",
"activityTitle": "**catta202000**",
"activityImage": "https://assets-cdn.github.com/images/email/message_cards/avatar.png",
"activitySubtitle": "@catta202000",
"facts": [

]
}
],
"potentialAction": [
{
"name": "Add a comment",
"@type": "ActionCard",
"inputs": [
{
"isMultiLine": true,
"@type": "TextInput",
"id": "IssueComment",
"isRequired": false
}
],
"actions": [
{
"name": "Comment",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueComment\",\n\"repositoryFullName\": \"mikacuy/pointnetvlad\",\n\"issueId\": 2,\n\"IssueComment\": \"{{IssueComment.value}}\"\n}"
}
]
},
{
"name": "Close issue",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueClose\",\n\"repositoryFullName\": \"mikacuy/pointnetvlad\",\n\"issueId\": 2\n}"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/mikacuy/pointnetvlad/issues/2#issuecomment-395726277"
}
],
"@type": "OpenUri",
"name": "View on GitHub"
},
{
"name": "Unsubscribe",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"MuteNotification\",\n\"threadId\": 342480803\n}"
}
],
"themeColor": "26292E"
}

**Comment by cattaneod** [comment]
@SJTUsuperxu You can find my whole PyTorch implementation [here](https://github.com/catta202000/PointNetVlad-Pytorch)

**Comment by SJTUsuperxu** [comment]
Thanks a lot
